include/boost/url/impl/any_allocator.hpp

```diff
@@ -0,0 +1,211 @@
+        if (align <= N) {
+            using aligned_t = typename boost::type_with_alignment<N>::type;
+            using aligned_allocator_t = boost::allocator_rebind_t<Allocator, aligned_t>;
+            aligned_allocator_t(a).deallocate(reinterpret_cast<aligned_t*>(p), n * ceil_div(size, align));
+        } else {
+            try_deallocate_aligned<2*N>(a, p, n, size, align);
+        }
+    }
+
+    template <class Allocator>
+    void deallocate_aligned(const Allocator& a, char* p, std::size_t n, std::size_t size, std::size_t align) {
+        try_deallocate_aligned<1>(a, p, n, size, align);
+    }
+}
+
+template <class T>
+template <class Allocator>
+any_allocator<T>::
+any_allocator(Allocator const& a) noexcept
+{
+    class impl;
+
+    using char_allocator_type = boost::allocator_rebind_t<Allocator, char>;
+    using impl_allocator_type = boost::allocator_rebind_t<Allocator, impl>;
+
+    class impl : public detail::any_allocator_base
+    {
+         char_allocator_type a_;
+
+    public:
+        explicit
+        impl(char_allocator_type const& a)
+        : any_allocator_base(typeid(char_allocator_type)), a_(a)
+        {
+        }
+
+        char*
+        allocate(std::size_t n, std::size_t size, std::size_t align) override
+        {
+            if (size == 1) {
+                return a_.allocate(n);
+            } else {
+                return detail::allocate_aligned(a_, n, size, align);
+            }
+        }
+
+        void
+        deallocate(char* p, std::size_t n, std::size_t size, std::size_t align)
+        {
+            if (size == 1) {
+                a_.deallocate(p, n);
+            } else {
+                detail::deallocate_aligned(a_, p, n, size, align);
+            }
+        }
+
+        bool
+        is_equal(const any_allocator_base& other) override {
+            const impl* other_impl = dynamic_cast<const impl*>(&other);
+            if (!other_impl) {
+                return false;
+            }
+            return a_ == other_impl->a_;
+        };
+    };
+
+    impl_allocator_type al(a);
+    p_ = std::allocate_shared<impl>(al, char_allocator_type{});
+}
+
+template <class T>
+any_allocator<T>::
+any_allocator(
+    any_allocator const& other) noexcept : p_(other.p_)
+{}
+
+template <class T>
+template <class U>
+any_allocator<T>::
+any_allocator(
+    any_allocator<U> const& other) noexcept
+: p_(other.p_)
+{}
+
+template <class T>
+any_allocator<T>&
+any_allocator<T>::
+operator=(
+    any_allocator<T> const& other) noexcept
+{
+    p_ = other.p_;
+    return *this;
+}
+
+template <class T>
+any_allocator<T>::
+any_allocator(
+    any_allocator&& other) noexcept : p_(std::move(other.p_))
+{
+    other.p_ = nullptr;
+}
+
+template <class T>
+template <class U>
+any_allocator<T>::
+any_allocator(
+    any_allocator<U> && other) noexcept
+: p_(other.p_)
+{
+    other.p_ = nullptr;
+}
+
+template <class T>
+T*
+any_allocator<T>::
+allocate(
+    std::size_t n) const
+{
+    if (p_ == nullptr)
```

Comment:
write the condition so that the most common case evaluates to true. Default-constructed allocators are rare, so the expression should read `if( p_ != nullptr )`